tests/deploy/test_scheduler_agent.py

```diff
@@ -1816,15 +1816,303 @@ async def test_scheduler_priority(agent: TestAgent, environment, make_resource_m
     second_task = await agent.scheduler._work.agent_queues.queue_get("agent1")
-    assert isinstance(second_task[0], tasks.DryRun)
+    assert isinstance(second_task.task, tasks.DryRun)
 
     # All tasks were consumed
     queue = agent.scheduler._work.agent_queues._get_queue("agent1")._queue
     assert len(queue) == 0
 
 
+async def test_scheduler_priority_rescheduling(agent: TestAgent, environment, make_resource_minimal):
+    """
+    Verify behavior of task rescheduling (e.g. because of new requires) with respect to priorities, both explicit and relative
+    to other scheduled task (i.e. current queue position).
+    """
+
+    # General flow of the test: all deploy tasks, all on the same agent, different priorities. Tasks are never picked up,
+    # only added or shifted around (due to changing priorities or requires edges).
+    # This test interacts with scheduled work directly rather than to go through the scheduler for verbosity reasons.
+    # For a basic priority test that does go via the scheduler see test_scheduler_priority.
+
+    # define some resources
+    rid1 = ResourceIdStr("test::Resource[agent,name=1]")
+    rid2 = ResourceIdStr("test::Resource[agent,name=2]")
+    rid3 = ResourceIdStr("test::Resource[agent,name=3]")
+    rid4 = ResourceIdStr("test::Resource[agent,name=4]")
+    rid5 = ResourceIdStr("test::Resource[agent,name=5]")
+    rid6 = ResourceIdStr("test::Resource[agent,name=6]")
+    task1 = tasks.Deploy(resource=rid1)
+    task2 = tasks.Deploy(resource=rid2)
+    task3 = tasks.Deploy(resource=rid3)
+    task4 = tasks.Deploy(resource=rid4)
+    task5 = tasks.Deploy(resource=rid5)
+    task6 = tasks.Deploy(resource=rid6)
+
+    # set up the scheduled work
+    requires: state.RequiresProvidesMapping = state.RequiresProvidesMapping()
+    work: ScheduledWork = ScheduledWork(
+        requires=requires, provides=requires.provides_view(), new_agent_notify=lambda *args, **kwargs: None
+    )
+
+    # three different deploy requests, each for a different resource. Verify that order ends up as expected
+    work.deploy_with_context(
+        resources={rid1},
+        reason="First deploy",
+        priority=TaskPriority(2),
+    )
+    work.deploy_with_context(
+        resources={rid2},
+        reason="Second deploy",
+        priority=TaskPriority(1),
+    )
+    work.deploy_with_context(
+        resources={rid3},
+        reason="Third deploy",
+        priority=TaskPriority(3),
+    )
+
+    assert [item.task for item in work.agent_queues.sorted("agent")] == [task2, task1, task3]
+
+    # add some more tasks to the middle priority level
+    work.deploy_with_context(
+        resources={rid1, rid2, rid3, rid4, rid5},
+        reason="Bulk deploy",
+        priority=TaskPriority(2),
+    )
+    # add one more task to the more urgent priority level for some relative ordering asserts later on
+    # -> priority 1: rid2, rid6
+    # -> priority 2: rid1, rid3, rid4, rid5 (latter three in any order)
```

Comment:
I wonder if we want to always make the tie breaker unique?

Reply:
We could address it in this test case by requesting three separate deploys instead of the one bulk deploy, but I think that would add verbosity for little gain (we don't need to know the order, we only need to inspect it so we can assert that it is maintained further on).